File: Army/ability/priestability.cpp
#include "priestability.h"

PriestAbility::PriestAbility(Unit* currentUnit) : Ability(currentUnit) {}

PriestAbility::~PriestAbility() {}

void PriestAbility::castSpell(Unit* victim, Spell* spell) {
    if ( spell->getSpellsType() == battleSpell ) {
        if ( victim->getIsDead() ) {
        victim->takeMagicDamage(spell->getHitPoints()*2);

        return;
        }
        victim->takeMagicDamage(spell->getHitPoints()/2);

        return;
    }
    if ( spell->getSpellsType() == healSpell ) {
        victim->heal(spell->getHitPoints()*1.5);

        return;
    }
    if ( spell->getSpellsType() == blessedSpell ) {
        ensureIsNotInfected(victim);
        victim->setNextState(new AngelState(victim));

        return;
    }
}

void PriestAbility::ensureIsNotInfected(Unit* victim) {
    if ( victim->getUnitType() == werewolf || victim->getUnitType() == vampire ) {
        throw IsInfectedUnitsException();
    }

}

File: Army-Java/Ability/PriestAbility.java
package com.gymfox.Army.Ability;

import com.gymfox.Army.Spells.Spell;
import com.gymfox.Army.Units.Unit;

public class PriestAbility extends Ability {
    public PriestAbility(Unit currentUnit) {
        super(currentUnit);
    }

    public void castSpell(Unit victim, Spell currentSpell) throws Unit.UnitIsDeadException {
        if ( currentSpell.getSpellsType() == Spell.SpellsType.BATTLESPELL ) {
            if ( victim.getIsDead() ) {
                victim.takeMagicDamage(currentSpell.getHitPoints() * 2);

                return;
            }
            victim.takeMagicDamage(currentSpell.getHitPoints() / 2);

            return;
        }
        if ( currentSpell.getSpellsType() == Spell.SpellsType.HEALSPELL) {
            victim.heal(currentSpell.getHitPoints());

            return;
        }
    }
}
